### PR TITLE
bootrr-generic-tests: Add TPM boot sanity test

### PR DESCRIPTION
**Pull request by 10ne1**
This adds a new simple smoke test to verify TPM chip
communication by querrying any commands it exposes via
an implementation provided by the tpm2-tools project.

More in-depth TPM functionality testing than this smoke
test should be done via the respective devices software,
for e.g. ChromeOS has its own custom userspace TPM stack
which can be exercised via the TAST test framework.

Signed-off-by: Adrian Ratiu <adrian.ratiu@collabora.com>

[Here is an example run](https://lava.collabora.co.uk/results/testcase/170445026) of this test on Octopus.

A separate pull request to enable the relevant kernel configs and install tpm2-tools in rootfs will be created soon.

### CODE REVIEW

**Comment by 10ne1** [comment]
FYI: @gctucker & @eballetbo I'm pinging you here because I cannot add you as explicit reviewers on this pull request

**Comment by eballetbo** [comment]
bootrr is meant to be as simple as possible. I'm wondering if this test could be a bit more simple just checking the caps property in sysfs is present. E.g:

`cat /sys/class/tpm/tpm0/caps`

If the kernel shows this property some communication between the kernel and the tpm chip already happened so basically is a similar test. FWIW, then you don't need to install tpm2-tools in the rootfs.

As a complement for this patch would be enable or check if all the tpm chips are enabled by default for the different Chromebooks, maybe we should add some TPM configs on some of them.

**Comment by 10ne1** [comment]
@eballetbo I thought about that but unfortunately not all drivers expose the same `/sys/class/tpm/` information, for example the cr50 i2c driver on Chromebook Eve does not expose caps:

```
ls /sys/class/tpm/tpm0/caps
ls: cannot access '/sys/class/tpm/tpm0/caps': No such file or directory
```

The only way which should work with all TPM chips is to use tpm2-tools to open and validate communication directly with the /dev/tpm device. The size of tpm2-tools is negligible anyway and the code complexity is just a one-liner shell call :)

**Comment by eballetbo** [comment]
Far enough. Then this change LGTM

File: helpers/assert_tpm_bringup [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+. bootrr
+
+TEST_CASE_ID="$1"
+TIMEOUT="${2:-1}"
```
Comment (gctucker) [inline helpers/assert_tpm_bringup]: 1 second might be a bit short for a timeout.  It's probably wiser to have 10s timeout and reduce the risk of having false positive due to random factors that might slow the system down temporarily.

Comment (10ne1) [reply helpers/assert_tpm_bringup]: Done. Timeout is now 10s.

**Comment by gctucker** [comment]
@10ne1 The test got run on staging:
https://staging.kernelci.org/test/case/id/616699b03de5adf519a6c459/

However, as you can see in the log fragment, it's failing as it doesn't seem to like the command line:
```
tpm2_getcap: invalid option -- 'c'
```

**Comment by 10ne1** [comment]
> @10ne1 The test got run on staging: https://staging.kernelci.org/test/case/id/616699b03de5adf519a6c459/
> 
> However, as you can see in the log fragment, it's failing as it doesn't seem to like the command line:
> 
> ```
> tpm2_getcap: invalid option -- 'c'
> ```

So Debian ships tpm2-tools v3 and buildroot ships a newer version. The `-c` option got removed in v4. v5 is the current one. If there is no way to install a newer tpm2-tools in debian then I'll have to test the existence of `-c` because tpm2-tools has no cross-distro way of testing the version (eg something like tpm2_getcap --version).

**Comment by 10ne1** [comment]
Updated the review to use the v4 + style and will remove tpm2-tools from Debian. If necessary we could add it back after upgrading to the latest Bullseye stable release.

**Comment by gctucker** [comment]
> > @10ne1 The test got run on staging: https://staging.kernelci.org/test/case/id/616699b03de5adf519a6c459/
> > However, as you can see in the log fragment, it's failing as it doesn't seem to like the command line:
> > ```
> > tpm2_getcap: invalid option -- 'c'
> > ```
> 
> So Debian ships tpm2-tools v3 and buildroot ships a newer version. The `-c` option got removed in v4. v5 is the current one. If there is no way to install a newer tpm2-tools in debian then I'll have to test the existence of `-c` because tpm2-tools has no cross-distro way of testing the version (eg something like tpm2_getcap --version).

No, the error is with buildroot.  The bootrr test suite is not run with Debian anywhere in KernelCI at the moment.

**Comment by 10ne1** [comment]
> > > @10ne1 The test got run on staging: https://staging.kernelci.org/test/case/id/616699b03de5adf519a6c459/
> > > However, as you can see in the log fragment, it's failing as it doesn't seem to like the command line:
> > > ```
> > > tpm2_getcap: invalid option -- 'c'
> > > ```
> > 
> > 
> > So Debian ships tpm2-tools v3 and buildroot ships a newer version. The `-c` option got removed in v4. v5 is the current one. If there is no way to install a newer tpm2-tools in debian then I'll have to test the existence of `-c` because tpm2-tools has no cross-distro way of testing the version (eg something like tpm2_getcap --version).
> 
> No, the error is with buildroot. The bootrr test suite is not run with Debian anywhere in KernelCI at the moment.

Yes, the error is with buildroot because we were passing it an obsolete flag which only exists in debian `-c`. It is fixed now, I removed the flag.

**Review by gctucker — APPROVED** [review]
Tested OK with staging build on grunt (finally):
https://lava.collabora.co.uk/scheduler/job/5674285

**Pull request merged**